enhancements/cloud-integration/openstack/failure-domains.md

```diff
@@ -0,0 +1,720 @@
+          image: rhcos
+          failureDomain:
+            computeAvailabilityZone: az1
+            ports:
+            - networkID: 4a270cf8-f5f9-4a09-8cfc-d39d77aa2ba1
+              fixedIPs:
+              - subnetID: 31ad4088-7e52-48da-9df5-35e70da438f6
+              tags:
+              - "control-plane"
+```
+
+#### Variation [optional]
+
+A failure domain with only compute availability zones:
+
+```yaml
+failureDomain:
+  computeAvailabilityZone: az1
+```
+
+A failure domain with only storage availability zones:
+
+```yaml
+failureDomain:
+  storageAvailabilityZone: az1
+```
+
+A failure domain with only network ports:
+
+```yaml
+failureDomain:
+  ports:
+  - networkID: 4a270cf8-f5f9-4a09-8cfc-d39d77aa2ba1
+    fixedIPs:
+    - subnetID: 31ad4088-7e52-48da-9df5-35e70da438f6
+    tags:
+    - "control-plane"
+```
+
+A failure domain defining both control plane and storage network ports:
+
+```yaml
+failureDomain:
+  ports:
+  - networkID: 4a270cf8-f5f9-4a09-8cfc-d39d77aa2ba1
+    fixedIPs:
+    - subnetID: 31ad4088-7e52-48da-9df5-35e70da438f6
+    tags:
+    - "control-plane"
+  - networkID: deb137b1-8ea4-4674-ab46-4cdaaa27637d
+    fixedIPs:
+    - subnetID: e5cb5138-8427-4cbe-ac10-bcf86b7b0db0
+    tags:
+    - "storage"
+```
+
+An empty failure domain is valid, but probably not very useful:
+
+```yaml
+failureDomain:
+```
+
+### API Extensions
+
+We change 2 APIs:
+
+* [OpenStack `MachinePool`](https://github.com/openshift/installer/blob/dbbc890fa40bd49aa761fb03612415e964c9eaf2/pkg/types/openstack/machinepool.go#L5-L35) as described in [the Installer section](#installer).
+* [`OpenStackProviderSpec`](https://github.com/openshift/api/blob/54592eea55395af31e82fa9a605eb45523115454/machine/v1alpha1/types_openstack.go#L30-L107) as described in [the `OpenStackProviderSpec` section](#openstackproviderspec).
+
+In both cases we are strictly adding new functionality rather than modifying existing behaviour.
+
+OpenStack `MachinePool` is referenced in both the `compute` and `controlPlane` sections of install-config with behaviour defined above.
+
+`OpenStackProviderSpec` is not strictly directly referenced by any CRD as it is only ever embedded as a Raw value. However, being embedded in the `MachineSpec` it is effectively referenced by `Machine` and `MachineSet`. It does not modify the behaviour of any component other than `machine-api-provider-openstack` (the OpenStack machine controller) as described above.
+
+Integration with Control Plane MachineSet will require a modification to that API. However, while it is planned it is not yet defined.
+
+### Implementation Details/Notes/Constraints
+
+We will implement this behaviour in upstream CAPO first. Critically the points of similarity will be:
+
+* The OpenStackFailureDomain struct will be identical in OpenShift and CAPO. To ensure the independence of the OpenShift API we will have a copy rather than a reference, but the intention will be for them to remain in lock step.
```

Comment:
Does this make sense as a goal? OpenShift must maintain the existing APIs we have while the upstream CAPO APIs are still evolving and can make breaking changes. I would argue that as long as we can reliably convert between the v1alpha1 OpenStack provider spec and the CAPO struct, then we are compatible.

And it may be advantageous, from an OCP perspective, to not duplicate the design exactly, for example, it may confusing in the context of the existing provider spec